### PR TITLE
Implement JavaScript major mode

### PR DESCRIPTION
**Pull request by nodeselector**
First iteration of a JavaScript major mode layer (see #98). Currently missing REPL capabilities because I don't think any of the REPL extensions for javascript are useful, but if someone has a REPL they like I'm happy to add it now or later.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2844,6 +2844,278 @@
...
+											{
+												"key": "t",
+												"name": "Tasks",
+												"type": "bindings",
+												"bindings": [
+													{
+														"key": "n",
+														"name": "npm",
+														"type": "command",
+														"command": "npm-script.run"
```
Comment (stevenguh) [inline package.json:3101]: Would this be the same as the built-in command `npm.runScript`?

File: package.json [hunk 2]
```diff
@@ -2844,6 +2844,278 @@
...
+																"key": "t",
+																"name": "convert string concat to template literal",
+																"type": "command",
+																"command": "cmstead.jsRefactor.convertToTemplateLiteral"
+															}
+														]
+													},
+													{
+														"key": "w",
+														"name": "wrap",
```
Comment (stevenguh) [inline package.json:3024]:
```suggestion
														"name": "+Wrap",
```

Discussion: should the name be surround like the vim?

File: package.json [hunk 3]
```diff
@@ -2844,6 +2844,278 @@
...
+												]
+											},
+											{
+												"key": "r",
+												"name": "+Refactor",
+												"type": "bindings",
+												"bindings": [
+													{
+														"key": "a",
+														"name": "mark function as async",
```
Comment (stevenguh) [inline package.json:2939]: Capitalized the first letter all the names (applies to all other bindings as well) to be consistent

File: package.json [hunk 4]
```diff
@@ -2844,6 +2844,278 @@
...
+														"type": "command",
+														"command": "cmstead.jsRefactor.markAsAsync"
+													},
+													{
+														"key": "e",
+														"name": "export",
```
Comment (stevenguh) [inline package.json:2945]:
```suggestion
														"name": "Add export",
```

missing the word Add?

**Review by stevenguh — CHANGES_REQUESTED** [review]
Thanks for submitting the change to add JS/TS.  It looks good overall 😊 Since the JS is the same as the typescript, I will only comment on this PR.

Can you document what extensions are required like https://github.com/VSpaceCode/vspacecode.github.io/pull/42?

**Comment by MarcoIeni** [comment]
Is not possible to use [conditional operators](https://code.visualstudio.com/docs/getstarted/keybindings#_conditional-operators) to assign the same key bindings to javascript and typescript right?
Should I open a feature request in which key?

**Comment by stevenguh** [comment]
> Is not possible to use [conditional operators](https://code.visualstudio.com/docs/getstarted/keybindings#_conditional-operators) to assign the same key bindings to javascript and typescript right?
> Should I open a feature request in which key?

Currently. It's not possible, but yeah feel free to open an issue :) I also have been thinking how to change the way major mode and condition work together.

File: package.json (hunk 2 above)
Comment (nodeselector) [reply package.json:3024]: I think I prefer surround myself now that you suggest it.

File: package.json [hunk 5]
```diff
@@ -2844,6 +2844,253 @@
...
+															{
+																"key": "t",
+																"name": "Convert string concat to template literal",
+																"type": "command",
+																"command": "cmstead.jsRefactor.convertToTemplateLiteral"
+															}
+														]
+													},
+													{
+														"key": "w",
```
Comment (stevenguh) [inline package.json:3004]:
```suggestion
														"key": "s",
```
Should the key be `s` then?

Comment (nodeselector) [reply package.json]: I ended up dropping this, [unfortunately](https://github.com/VSpaceCode/vspacecode.github.io/pull/43).

File: package.json [hunk 6]
```diff
@@ -2844,6 +2844,140 @@
...
+													{
+														"key": "r",
+														"name": "Rename symbol",
+														"type": "command",
+														"command": "editor.action.rename"
+													}
+												]
+											},
+											{
+												"key": "t",
```
Comment (stevenguh) [inline package.json:2961]:
```suggestion
												"key": "n",
```
Should this be `n` like spacemacs' JS layer?

**Comment by MarcoIeni** [comment]
> Currently. It's not possible, but yeah feel free to open an issue :) I also have been thinking how to change the way major mode and condition work together.

I found out that what I wanted in order to remove code duplication from this kind of situations (and major modes in general) is something like [this](https://github.com/VSpaceCode/vscode-which-key/issues/13#issuecomment-678077861), so I will not open a new issue

**Comment by stevenguh** [comment]
> I found out that what I wanted in order to remove code duplication from this kind of situations (and major modes in general) is something like [this](https://github.com/VSpaceCode/vscode-which-key/issues/13#issuecomment-678077861), so I will not open a new issue

I opened https://github.com/VSpaceCode/vscode-which-key/issues/42 because the up-coming release v0.9.0 only implemented menu declarations for transient menu only (since I broke out transient as a separate command).

**Comment by The-Compiler** [comment]
What's the state of this and #1094? I'd love to see this as well, and would be happy to help introducing icons (see #205) either before or after this is merged.

**Comment by stevenguh** [comment]
I think there are 3 things are needed
- Commit the suggested changes
- Add icons
- Duplicate that to TS for now

The duplication can be remove once we implemented
- https://github.com/VSpaceCode/vscode-which-key/issues/42
- https://github.com/VSpaceCode/VSpaceCode/issues/199

@The-Compiler Wondering if you are willing add icons? I can commit the suggestion and merge this, then you can work to add icons as a separate PR? At the end, I can duplicate the JS binding for TS?

**Comment by MarcoIeni** [comment]
I think we can duplicate the bindings for the moment for js and ts, yes

**Comment by stevenguh** [comment]
Close this due to stale PR. I just added basic js/ts bindings @ c6f08b352a828ae96599c140cded648385a6ef59.

**Pull request closed without merging**